modules/standard/Lists.chpl

```diff
@@ -0,0 +1,907 @@
+
+      :return: The index of the element to search for.
+
+      :throws IllegalArgumentError: If the given element cannot be found.
+    */
+    proc indexOf(x: eltType, start: int=1, end: int=size): int throws {
+      _enter();
+
+      try _boundsCheckLeaveOnThrow(start);
+      try _boundsCheckLeaveOnThrow(end);
+
+      for i in start..end do
+        if x == _getRef(i) {
+          _leave();
+          return i;
+        }
+
+      _leave();
+
+      const msg = "No such element in list: " + x:string;
+      throw new owned
+        IllegalArgumentError(msg);
+
+      // Should never reach here.
+      return -1;
+    }
+
+    /*
+      Return the number of times a given element is found in this list.
+
+      :arg x: An element to count.
+
+      :return: The number of times a given element is found in this list.
+      :rtype: `int`
+    */
+    proc count(x: eltType): int {
+      _enter();
+
+      var result = 0;
+
+      for i in 1.._size do
+        if x == _getRef(i) then
+          result += 1;
+
+      _leave();
+      return result;
+    }
+
+    /*
+      Sort the items of this list in place using a comparator. If no comparator
+      is provided, sort this list using the default sort order of its elements.
+
+      .. warning::
+
+        Sorting the elements of this list may invalidate existing references
+        to the elements contained in this list.
+
+      :arg comparator: A comparator used to sort this list.
+    */
+    proc sort(comparator=Sort.defaultComparator) {
+      //
+      // TODO: This is not ideal, but the Sort API needs to be adjusted before
+      // we can sort over lists directly.
+      //
+      var array = toArray();
+      clear();
+      Sort.sort(array, comparator);
+      extend(array);
+    }
+
+    /*
+      Index this list via subscript. Returns a reference to the element at a
+      given index in this list.
+
+      :arg i: The index of the element to access.
+
+      .. warning::
+
+        Use of the `this` method with an out of bounds index (while bounds
+        checking is on) will cause the currently running program to halt.
+
+      :return: An element from this list.
+    */
+    proc const this(i: int) ref {
+      _enter();
+
+      if boundsChecking && !_withinBounds(i) {
+        _leave();
+        const msg = "Invalid list index: " + i:string;
+        halt(msg);
+      }
+
+      ref result = _getRef(i);
+      _leave();
+      return result;
+    }
+
+    /*
+      Produce a serial iterator over the elements of this list.
+
+      :yields: A reference to one of the elements contained in this list.
+    */
+    iter these() ref {
+
+      //
+      // TODO: I'm not even sure of what the best way to WRITE a threadsafe
+      // iterator is, _let alone_ whether it should even be threadsafe
+      // (I mean, is there even a point when reference/iterator invalidation
+      // is still a thing?).
+      //
+      for i in 1.._size {
+        _enter();
```

Comment:
Isn't _size protected by your lock, though? I'm not so confident this is correct. Consider creating a TODO issue to dig into the topic and answer the above questions.